coderd/workspaces.go

```diff
@@ -103,38 +103,60 @@ func (api *API) workspace(rw http.ResponseWriter, r *http.Request) {
 func (api *API) workspaces(rw http.ResponseWriter, r *http.Request) {
 	apiKey := httpmw.APIKey(r)
 
-	// Empty strings mean no filter
-	orgFilter := r.URL.Query().Get("organization_id")
-	ownerFilter := r.URL.Query().Get("owner")
-	nameFilter := r.URL.Query().Get("name")
-
-	filter := database.GetWorkspacesWithFilterParams{Deleted: false}
-	if orgFilter != "" {
-		orgID, err := uuid.Parse(orgFilter)
-		if err == nil {
-			filter.OrganizationID = orgID
-		}
+	queryStr := r.URL.Query().Get("q")
+	values, err := WorkspaceSearchQuery(queryStr)
+	if err != nil {
+		httpapi.Write(rw, http.StatusBadRequest, httpapi.Response{
+			Message: "Invalid workspace search query.",
+			Validations: []httpapi.Error{
+				{Field: "q", Detail: err.Error()},
+			},
+		})
+		return
 	}
-	if ownerFilter == "me" {
-		filter.OwnerID = apiKey.UserID
-	} else if ownerFilter != "" {
-		userID, err := uuid.Parse(ownerFilter)
-		if err != nil {
-			// Maybe it's a username
-			user, err := api.Database.GetUserByEmailOrUsername(r.Context(), database.GetUserByEmailOrUsernameParams{
-				// Why not just accept 1 arg and use it for both in the sql?
-				Username: ownerFilter,
-				Email:    ownerFilter,
+
+	// Set all the query params from the "q" field.
+	q := r.URL.Query()
+	for k, v := range values {
+		// Do not allow overriding if the user also set query param fields
+		// outside the query string.
+		if q.Has(k) {
+			httpapi.Write(rw, http.StatusBadRequest, httpapi.Response{
+				Message: fmt.Sprintf("Workspace filter %q cannot be set twice. In query params %q and %q.", k, k, "q"),
 			})
-			if err == nil {
-				filter.OwnerID = user.ID
-			}
-		} else {
-			filter.OwnerID = userID
+			return
 		}
+		q.Set(k, v)
 	}
-	if nameFilter != "" {
-		filter.Name = nameFilter
+	r.URL.RawQuery = q.Encode()
+
+	parser := httpapi.NewQueryParamParser()
+	filter := database.GetWorkspacesWithFilterParams{
+		Deleted:        false,
+		OrganizationID: parser.ParseUUID(r, uuid.Nil, "organization_id"),
+		OwnerID:        parser.ParseUUIDorMe(r, uuid.Nil, apiKey.UserID, "owner_id"),
+		OwnerUsername:  parser.ParseString(r, "", "owner"),
+		TemplateName:   parser.ParseString(r, "", "template"),
+		TemplateIds:    parser.ParseUUIDArray(r, []uuid.UUID{}, "template_ids"),
+		Name:           parser.ParseString(r, "", "name"),
+	}
+	if len(parser.ValidationErrors()) > 0 {
+		httpapi.Write(rw, http.StatusBadRequest, httpapi.Response{
+			Message:     "Query parameters have invalid values.",
+			Validations: parser.ValidationErrors(),
+		})
+		return
+	}
+
+	if filter.OwnerUsername == "me" {
+		if !(filter.OwnerID == uuid.Nil || filter.OwnerID == apiKey.UserID) {
+			httpapi.Write(rw, http.StatusBadRequest, httpapi.Response{
+				Message: "Cannot set both \"me\" in \"owner_name\" and use \"owner_id\".",
+			})
+			return
+		}
```

Comment:
I'm kinda hesitant to return 4xx errors in this route since it's powered by a filter. We can add support for validation error in the query bar on the dashboard, but right now we the behavior is we just take the first. 

Could we just drop ownerID? I think we use owner name on the app everywhere anyways?

Reply:
I'm ok with dropping `owner_id` for now :+1:

Reply:
I'm keeping it in the database filter because apiKey has userID, not their name. But I'll drop it from the query params.